### PR TITLE
Django-based cache-busting, integration environment and tests

### PR DESCRIPTION
**Pull request by coffeemancy**
This is an alternative to https://github.com/Anguirel86/ctjot_web_generator/pull/15 based on discussion there.

This approach uses Django's built-in [ManifestStaticFilesStorage](https://docs.djangoproject.com/en/4.1/ref/contrib/staticfiles/#django.contrib.staticfiles.storage.ManifestStaticFilesStorage) mechanism to automatically create hashed static files (javascript, CSS, images, etc.) and insert those for use in `{% static %}` templating calls.

In investigating that, it was revealed that this feature cannot work when DEBUG is set, like in dev (and currently in staging). Additionally, there was some concern about interaction with gunicorn/nginx.

This PR introduces a new environment: integration. This environment uses gunicorn and nginx like staging, but without SSL/letsencrypt nor wiki pages (similar to dev). This allows for local testing of the integration points of the django app, gunicorn, and nginx. Additionally this PR adds a [github workflow to spin up the integration environment and check all containers pass their healthchecks](https://github.com/coffeemancy/ctjot_web_generator/actions/runs/5898131095).

Lastly this adds a mechanism similar to the original PR for injecting query strings on static files with the file MD5 hashes for when DEBUG is set and ManifestStaticFilesStorage cannot be used.

## Updates

* Add [github workflow to run `shellcheck` on PRs](https://github.com/coffeemancy/ctjot_web_generator/actions/runs/5898131093), update scripts to pass
* Add integration testing environment deployment
* Add [github workflow for deploying and testing integration environment](https://github.com/coffeemancy/ctjot_web_generator/actions/runs/5898131095)
* Use ManifestStaticFilesStorage whenever DEBUG is not set
* Allow overriding DEBUG when running dev/integration environments
* Remove DEBUG from staging
* For dev or when DEBUG is on, use query-based cache-busting for static files via monkeypatching the `{% static %}` template tag
* Prefer "docker compose" over "docker-compose" when available and check that docker and compose are installed in `deploy.sh`
* Add healthchecks to containers
* Use [`extends`](https://docs.docker.com/compose/multiple-compose-files/extends/) so compose files all inherit from prod one and override which env vars files they use

### Dev Env

Ran `./deploy/deploy.sh -d` and can see that the query strings are inserted into static files requests:

screenshot-static-files-dev

### Integration

Ran `./deploy/deploy.sh -i` and can see that the hashed static files are used:

screenshot-static-files

## Notes

This removes DEBUG from staging, so can get ManifestStaticFilesStorage to be used, which may or may not be desired(?).

### CODE REVIEW

File: generator/apps.py [hunk 1]
```diff
@@ -1,6 +1,54 @@
...
     name = 'generator'
+
+    def __init__(self, app_name, app_module):
+        if bool(int(os.getenv("DEBUG", "0"))):
+            self._monkey_patch_static_node()
+
+        super(GeneratorConfig, self).__init__(app_name, app_module)
+
+    @staticmethod
+    def _monkey_patch_static_node():
```
Comment (Anguirel86) [inline generator/apps.py:23]: Clever.  I like that we don't have to modify the views/contexts with this approach.

File: deploy/.env.dev [hunk 2]
```diff
@@ -1,4 +1,3 @@
-DEBUG=1
```
Comment (Anguirel86) [inline deploy/.env.dev:1]: Is this being handled elsewhere now?

File: deploy/deploy.sh [hunk 3]
```diff
@@ -50,7 +84,7 @@ migrate_dokuwiki_data() {
     fi
 
     echo "Copying ${WIKIDATA}/data/${dir}..."
-    sudo cp -r ${WIKIDATA}/data/${dir}/* deploy/wiki_config/dokuwiki/data/${dir}/
+    sudo cp -r "${WIKIDATA}/data/${dir}/"* "deploy/wiki_config/dokuwiki/data/${dir}/"
```
Comment (Anguirel86) [inline deploy/deploy.sh:84]: Good catch.  I got kind of lazy with path quoting.

File: deploy/.env.dev (hunk 2 above)
Comment (coffeemancy) [reply deploy/.env.dev:1]: Yeah, in the compose file itself. It defaults to 1 in the `docker-compose.dev.yml` but you can override it with env var when you run it.

File: deploy/deploy.sh (hunk 3 above)
Comment (coffeemancy) [reply deploy/deploy.sh:84]: This was thanks to `shellcheck` :)

**Comment by Anguirel86** [comment]
I'm a little torn on the dev vs integration vs staging deployments.  Just sort of thinking out loud (in text?) here and using PR comments as my rubber duck:

The original idea behind the three deployment types was that dev is used for quickly testing local changes.  Staging is supposed to be the final checkout before we deploy live, just to make sure that everything works in the fully containerized environment, and prod is pretty self explanatory.  

Losing the DEBUG flag by default in staging is undesirable, since any Django issues will fail with no real details.  Though we definitely want DEBUG off for staging now that we have different cache busting strategies depending on that flag.  I guess we can just enable debug mode temporarily if Django starts throwing out 500 error codes in staging.  

So the integration deployment adds nginx, gunicorn, and DEBUG off (compared against dev), and we lose wiki and SSL from staging.  Wiki isn't an issue. It's only really ever used on the live server anyway.  SSL is mostly just a sanity check in staging to make sure I didn't fat finger something in the env files before the push to prod.  

Is it possible to use the staging staging (docker-compose and env files) for integration, but only start up a subset of the services?  ie: `docker-compose -f deploy/docker-compose.staging.yml up --build -d web-generator db nginx-proxy`?  I'm not sure how messy that would get with environment though, since the reverse proxy gets the host names from the environment.

Thoughts?  I'm not even sure I had a real question in that rambling mess.  I'm mostly just trying to think how we can keep things as simple as possible, and I may be overthinking things.

File: deploy/.env.dev (hunk 2 above)
Comment (Anguirel86) [reply deploy/.env.dev:1]: Ah, there it is.  Thanks.

**Comment by coffeemancy** [comment]
Based on some discord conversation, going to look at using https://docs.docker.com/compose/profiles/ to set up profiles and combine the dev and integration into single compose file / environment files, with gunicorn/nginx off by default (so same flow as current dev env).

Then update deploy.sh so "deploy.sh -d" still stays as-is (no gunicorn nor nginx) but "deploy.sh -i" adds them in for integration testing (which should be rare, only when mucking with non-DEBUG django internals, gunicorn, or nginx stuff).

**Comment by coffeemancy** [comment]
Turns out `--profile` for compose are bit weird.

You then have to explicitly always pass `--profile` to docker-compose (or set `COMPOSE_PROFILES` env var) for every command. So if you're used to running like `docker-compose ps` and such, it can interrupt  you (have to remember to set every time). You also seem to need to explicitly do a build before `up` or it skips it? It also doesn't start it's dependencies.

So I've attempted an alternative here, using `extends`: https://docs.docker.com/compose/multiple-compose-files/extends/

Now, staging, int, and dev all extend from prod compose file, inheriting all the settings there for each service ([except for `depends_on` apparently](https://docs.docker.com/compose/multiple-compose-files/extends/#exceptions-and-limitations)). So they'll always use same 3rd party images (like for db), settings, healthchecks, etc., and you can just override what you want in the individual compose files. So like staging overrides to use `.env.staging`, int and dev use `.env.dev`, etc.

I also added healthchecks to some containers in prod (which gets inherited for other envs).

Hopefully this ends up in a more DRY solution and less chance of divergence or having to update multiple files in the future.

**Pull request merged**